lib/mcb/editor/base.rb

```diff
@@ -0,0 +1,31 @@
+module MCB
+  module Editor
+    class Base
+      attr_reader :provider
+
+      def initialize(provider:, requester:)
+        @provider = provider
+        @requester = requester
+
+        setup_cli
+        check_authorisation
+      end
+
+    protected
+
+      def setup_cli
+        raise NotImplementedError, "Needs to be implemented in Child Class"
+      end
+
+      def check_authorisation
+        raise NotImplementedError, "Needs to be implemented in Child Class"
+      end
+
+      def audit
+        Audited.audit_class.as_user(@requester) do
+          yield
+        end
+      end
```

Comment:
So, nice pattern yall, but why did we need an explicit mechanism like this when we have an existing implicit mechanism?

https://github.com/DFE-Digital/manage-courses-backend/blob/869ce4af455299a126f78015361d15c5404eb3a7/config/initializers/mcb_audit.rb#L7

Not being judgy, I realise there could be a reason this was missed or just didn't work.

Reply:
1. Didn't realise this file (`config/initializers/mcb_audit.rb`) existed
2. Used the same code (`Audited.store[:audited_user] = user`) and in test suite didn't work when testing Audited User
3. This uses the `requester` which is set in the Config so would always Audit.
4. When is `MCB_AUDIT_USER` set?

Reply:
> 2. Used the same code (Audited.store[:audited_user] = user) and in the test suite it didn't work when testing the Audited User

We have seen this work so not sure why it wouldn't.

> 3. This uses the requester which is loaded via the Config so would always Audit.

Sorry but not clear on what you're saying here. Don't we always want to audit? Doesn't the config initialiser achieve that?

> 4. When is MCB_AUDIT_USER set?

I should be set automatically by `init_rails` / `rails_runner`:

https://github.com/DFE-Digital/manage-courses-backend/blob/5a3daa0356a80254fc7fbc57dd68a35a5758c7c1/lib/mcb.rb#L35

Just to note, @faissalMT is following up on this, as should be obvious already. :)

Reply:
2. Had a method in `MCB::Editor::Base` that setup the Audited User and it worked on an individual test but not in the suite. So I had low confidence it would work in production.

3/4. Not knowing about the setting the ENV variable and initialiser just meant not clear what's going on. I didn't go hunting down the chain when responding last night.

If I remember correctly when using the MCB tool with @timabell audits were not stored with a User (when we were adding auditing to Grant/Revoke). So we didn't think any code existed that catered for this. Might be worthwhile double checking that current data has audits with the correct User recorded.

Reply:
This is probably why Audits were not happening:  https://github.com/DFE-Digital/manage-courses-backend/pull/733/commits/828d25d49e56c75170a4daa92c1d7147133125f1#diff-14af88d7d0f67b03757c0f66837db947L1